src/datagram.cpp

```diff
@@ -3,73 +3,257 @@
+            }
+            else
+            {
+                log::trace(log_cat, "Sending standard datagram...");
+                _last_dgram_id += 4;
+                send_buffer.emplace_back(_last_dgram_id, std::make_pair(data, std::move(keep_alive)));
+                _skip_next = true;
+            }
+
+            conn.packet_io_ready();
         });
     }
 
-    void DatagramIO::append_buffer(bstring_view buffer, std::shared_ptr<void> keep_alive)
+    std::vector<ngtcp2_vec> DatagramIO::pending()
     {
         log::trace(log_cat, "{} called", __PRETTY_FUNCTION__);
-        user_buffers.emplace_back(buffer, std::move(keep_alive));
 
-        conn.packet_io_ready();
+        std::vector<ngtcp2_vec> bufs{};
+        bufs.reserve(2);
+
+        auto& b = send_buffer.front();
+
+        auto& did = bufs.emplace_back();
+        did.base = const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(&b));
+        did.len = 2;
+
+        auto& dat = bufs.emplace_back();
+        dat.base = const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(b.second.first.data()));
+        dat.len = b.second.first.size();
+
+        return bufs;
     }
 
-    std::vector<ngtcp2_vec> DatagramIO::pending()
+    prepared_datagram DatagramIO::pending_datagram()
     {
         log::trace(log_cat, "{} called", __PRETTY_FUNCTION__);
 
-        std::vector<ngtcp2_vec> bufs{};
-        bufs.reserve(std::distance(user_buffers.begin(), user_buffers.end()));
+        std::pair<uint64_t, std::vector<ngtcp2_vec>> bufs;
+        bufs.second.reserve(2);
+
+        auto& b = send_buffer.front();
+
+        prepared_datagram d{};
+        d.id = b.first;
+        d.bufs_len = 1;
 
-        for (auto& b : user_buffers)
+        oxenc::write_host_as_big(static_cast<uint16_t>(b.first), d.dgid.data());
+
+        if (_packet_splitting)
         {
-            auto& temp = bufs.emplace_back();
+            d.bufs[0].base = d.dgid.data();
+            d.bufs[0].len = 2;
+            d.bufs_len++;
+        }
+
+        d.bufs[d.bufs_len - 1].base = const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(b.second.first.data()));
+        d.bufs[d.bufs_len - 1].len = b.second.first.size();
+
+        return d;
+    }
+
+    std::optional<bstring> DatagramIO::to_buffer(bstring_view data, uint16_t dgid)
+    {
+        log::trace(log_cat, "DatagramIO handed datagram with endian swapped ID: {}", dgid);
 
-            temp.base = const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(b.first.data()));
-            temp.len = b.first.size();
+        auto r = recv_buffer.receive(data, dgid);
+
+        if (r)
+            return r;
+
+        return std::nullopt;
+    }
+
+    std::optional<bstring> rotating_buffer::receive(bstring_view data, uint16_t dgid)
+    {
+        log::trace(log_cat, "{} called", __PRETTY_FUNCTION__);
+
+        {
+            std::lock_guard<std::mutex> buffer_lock(m);
```

Comment:
I *think* we can drop the mutex and atomics in here, if we ensure that we only ever touch them from the main thread.  (For the test suite I think that would mean we need to add some indirection that goes through a `call` rather than accessing things directly, but that's probably fine for the test suite).